token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs

```diff
@@ -81,27 +83,39 @@ makeTokenLimiter tokenLimiterConfig = do
 -- Warning: this function can block for long if there are other threads waiting to debit tokens.
 -- Use in conjunction with an Async.race timeout or similar if you want to avoid lengthy blocking.
 canDebit :: TokenLimiter -> Word64 -> IO Bool
-canDebit TokenLimiter {..} debit = withMVar tokenLimiterLastServiced $ \(lastServiced, countThen) -> do
-  now <- getMonotonicTimeNSec
-  let currentCount = computeCurrentCount tokenLimiterConfig lastServiced countThen now
-  let enoughAvailable = currentCount >= debit
-  pure enoughAvailable
+canDebit TokenLimiter {..} debit = do
+  -- If any thread is currently waiting, we definitely can't debit any tokens
+  -- currently.
+  currentlyWaiting <- isEmptyMVar tokenLimiterWaitLock
```

Comment:
> The problem is, the two MVar's are independent, so even if isEmptyMVar is False here, there isn't any guarantee that by the next line (withMVar) the tokenLimiterLastServiced MVar wouldn't already be taken.

This is a fair point.
In that case the output would still be correct but the call would still (sadly) block.

> If we want to stay with an MVar-only solution

Not necessarily. How would you suggest to do this?
I think you can't `threadDelay` in an `stm`...

Reply:
You can do using `unsafeIoToSTM`, and `threadDelay` is benign here, but a problem is that after the current waiter finishes, a random other blocked STM waiter would be restarted, without fairness guarantee. So starving a waiter is possible. I don't immediately see a way to solve that within STM (but it doesn't mean it can't).

Reply:
Hm, maybe separating the operation to two stages could work: first stage would be an STM action that would add a unique id (myThreadId?) to a STM queue. This would be performed with atomically. Second stage would then be an STM action that reads from the queue, and retries unless the read value is its own unique. This way there's an established FIFO order on the actions. The second stage could also do the accounting of the tokens, as these are all STM actions, they compose nicely.